url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/server/SOCServer.java

```diff
@@ -0,0 +1,9048 @@
+                    {
+                        if (ga.canDoMonopolyAction())
+                        {
+                            int[] monoPicks = ga.doMonopolyAction(mes.getResource());
+
+                            final String monoPlayerName = (String) c.getData();
+                            final String resName
+                                = " " + SOCResourceConstants.resName(mes.getResource()) + ".";
+                            String message = monoPlayerName + " monopolized" + resName;
+
+                            gameList.takeMonitorForGame(gaName);
+                            messageToGameExcept(gaName, c, new SOCGameTextMsg(gaName, SERVERNAME, message), false);
+
+                            /**
+                             * just send all the player's resource counts for the
+                             * monopolized resource
+                             */
+                            for (int i = 0; i < ga.maxPlayers; i++)
+                            {
+                                /**
+                                 * Note: This only works if SOCPlayerElement.CLAY == SOCResourceConstants.CLAY
+                                 */
+                                messageToGameWithMon(gaName, new SOCPlayerElement(gaName, i, SOCPlayerElement.SET, mes.getResource(), ga.getPlayer(i).getResources().getAmount(mes.getResource())));
+                            }
+                            gameList.releaseMonitorForGame(gaName);
+
+                            /**
+                             * now that monitor is released, notify the
+                             * victim(s) of resource amounts taken,
+                             * and tell the player how many they won.
+                             */
+                            int monoTotal = 0;
+                            for (int i = 0; i < ga.maxPlayers; i++)
+                            {
+                                int picked = monoPicks[i];
+                                if (picked == 0)
+                                    continue;
+                                monoTotal += picked;
+                                String viName = ga.getPlayer(i).getName();
+                                StringConnection viCon = getConnection(viName);
+                                if (viCon != null)
+                                    messageToPlayer(viCon, gaName,
+                                        monoPlayerName + "'s Monopoly took your " + picked + resName);
+                            }
+
+                            messageToPlayer(c, gaName, "You monopolized " + monoTotal + resName);
+                            sendGameState(ga);
+                        }
+                        else
+                        {
+                            messageToPlayer(c, gaName, "You can't do a Monopoly pick now.");
+                        }
+                    }
+                    else
+                   
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *bw*. The resource is referenced at lines 7717, 7722, 7725, 7728, 7731. The resource is closed at line 7739. There are other execution paths that do not contain closure statements, for example, when BufferedWriter.newLine throws an exception. Ensure *bw* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).